junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt

```diff
@@ -231,6 +231,71 @@ fun assertNotNull(actual: Any?, messageSupplier: () -> String) {
     Assertions.assertNotNull(actual, messageSupplier)
 }
 
+/**
+ * Example usage:
+ * ```kotlin
+ * val string: Any = ...
+ *
+ * assertInstanceOf<String>(string)
+ *
+ * // compiler smart casts string to a String object
+ * assertTrue(string.isNotEmpty())
+ * ```
+ * @see Assertions.assertInstanceOf
+ */
+@API(since = "5.11", status = EXPERIMENTAL)
+inline fun <reified T : Any> assertInstanceOf(actual: Any) {
```

Comment:
Initially I thought that there is no need to accept nullable values, as in Kotlin we can ensure that nullable value is passed in compile time. 
But it would be easier for developers if the method accepts nullable values, agree

Reply:
Nice. This is resolved. I would also add a nullable smart cast test to prevent regressions:
```kotlin
    @Test
    fun `assertInstanceOf with compiler nullable smart cast`() {
        val string: Any? = "string"
        assertInstanceOf<String>(string)
        assertFalse(string.isEmpty()) // smart cast to a non-null String object
    }
```